README.md
Updated version of Mark Grebe's [SDLTRS]:

  * Included all patches by [EtchedPixels]: banked memory support, Lowe LE18
  * Fixed various SegFaults: ROM Selection Menu, Scaling in Fullscreen
  * Reworked the TextGUI: new shortcuts and key bindings, help screen
  * Ctrl-A, Ctrl-C & Ctrl-V can now be used in the Emulator (CP/M & WordStar)
  * Display scanlines to simulate an old CRT monitor resolution
  * Access to real floppy disks works now on Linux
  * Tried to fix reported bugs to the original version
  * Port to SDL2 (see [BUILDING.md] and [SDL2])
  * Support Exatron Stringy Floppy for TRS-80 Model I
  * Select and execute CMD files directly in the Emulator
  * Save screenshot of the Emulator window as BMP file
  * Show Z80 registers in the window title bar
  * Adjust speed of Z80 CPU on the fly
  * Emulate Z80 memory refresh register
  * Support Holmes Sprinter II/III speed-up kits for TRS-80 Model I/III
  * Change Z80 CPU default MHz of each TRS-80 Model
  * More accurate emulation of Z80 block instructions
  * Joystick emulation with Mouse
  * Support Prologica CP-300/500 16kB ROM with extra 2kB Z80 monitor
  * Support Seatronics Super Speed-Up Board for all TRS-80 Models
  * Load and Save TRS-80 memory in debugger
  * Patch Model I ROM to boot from hard disk drive
  * Support EACA EG 3200 (Genie III) system
  * 480x192 HRG resolution for LNW80 and TCS SpeedMaster/Genie IIs
  * CP/M banking support for TRS-80 Model I clones
  * Support EG-64 Memory-Banking-Adaptor from TCS
  * Support Lubomir Soft Banker for TRS-80 Model I
  * Emulation of the TCS Genie IIIs system
  * Support EG 64.1 memory expansion for TRS-80 Model I
  * Support Schmidtke 80-Z Video Card for TRS-80 Model I
  * Emulate EG 3210 Programmable Graphics Adaptor for Genie III
  * Support some David Keil's TRS-80 Emulator extensions
  * Support Anitek MegaMem memory card for TRS-80 Model III and 4/4P
  * Emulate 4 MHz speedup hardware for TRS-80 Model III
  * Support TCS Genie IIs/SpeedMaster RAM 192 B memory expansion
  * Emulate 6845 CRTC Interlace Video Mode
  * Support VideoExtension HRG for EACA EG 3200 Genie III
  * Support Genieplus Memory Card for EACA EG 3200 Genie III
  * Support Prologica CP-500/M80 by Leonardo Brondani Schenkel
  * Partial emulation of Aster CT-80

SDL(2)TRS is based on Tim Mann's excellent TRS-80 emulator [xtrs] and also
has very low system requirements: it works on all platforms supported by the
[SDL] library, even on machines with only a few hundred MHz of CPU speed.

Browse the [Documentation] online and checkout [SDLTRS and the Video Genies]
by Fritz Chwolka ...

## License

  [BSD 2-Clause](LICENSE)

## Contributing
All contributions are welcome.

## Binaries

  * [sdltrs.exe]     (32-bit, needs [SDL.DLL] of [SDL-1.2.14] for Win9X)
  * [sdl2trs.exe]    (32-bit, needs [SDL2.DLL])
  * [sdl2trs64.exe]  (64-bit, needs [SDL2.DLL])

(Release 1.2.25 build with [MinGW] & [MinGW-w64])

## Packages

  * [sdltrs_1.2.25-1_i386.deb]    (32-bit, SDL)
  * [sdl2trs_1.2.25-1_i386.deb]   (32-bit, SDL2)
  * [sdltrs_1.2.25-1_amd64.deb]   (64-bit, SDL)
  * [sdl2trs_1.2.25-1_amd64.deb]  (64-bit, SDL2)

(Build on Debian 9/i386 & Linux Mint 21/amd64)

Thanks to Tércio Martins a package for Arch Linux is available in the [AUR].

## SDL2

The [SDL2] branch contains the SDL2 version with hardware rendering support.
SDL2 binaries and packages above are build on the [SDL2] branch.

The SDL2 version is available in [RetroPie] since version 4.6.6 and Valerio
Lupi's fork of [RetroPie-Setup] ...

## Forks

  * [SDLTRS-SH] with Mongoose web server for debugging by Sascha Häberling
  * [SDLTRS-TRS-IO] with integrated TRS-IO and FreHD by Arno Puder

## Screenshots

![screenshot](screenshots/sdltrs01.png)
![screenshot](screenshots/sdltrs02.png)
![screenshot](screenshots/sdltrs03.png)
![screenshot](screenshots/sdltrs04.png)
![screenshot](screenshots/sdltrs05.png)

[AUR]: https://aur.archlinux.org/packages/sdltrs
[BUILDING.md]: BUILDING.md
[Documentation]: https://jengun.gitlab.io/sdltrs
[EtchedPixels]: https://www.github.com/EtchedPixels/xtrs
[MinGW]: https://osdn.net/projects/mingw/
[MinGW-w64]: http://mingw-w64.org
[RetroPie]: https://github.com/RetroPie
[RetroPie-Setup]: https://github.com/valerino/RetroPie-Setup
[SDL]: https://www.libsdl.org
[SDL2]: https://gitlab.com/jengun/sdltrs/-/tree/sdl2
[SDL.DLL]: https://www.libsdl.org/download-1.2.php
[SDL2.DLL]: https://github.com/libsdl-org/SDL/releases/latest
[SDL-1.2.14]: https://www.libsdl.org/release/SDL-1.2.14-win32.zip
[SDLTRS]: http://sdltrs.sourceforge.net
[SDLTRS and the Video Genies]: http://www.myoldc.info/eaca_tcs_computer/sdltrs_and_the_videogenies.html
[SDLTRS-SH]: https://github.com/shaeberling/sdltrs
[SDLTRS-TRS-IO]: https://github.com/apuder/sdltrs-trs-io
[sdltrs.exe]: bin/sdltrs.exe
[sdl2trs.exe]: bin/sdl2trs.exe
[sdl2trs64.exe]: bin/sdl2trs64.exe
[sdltrs_1.2.25-1_i386.deb]: bin/sdltrs_1.2.25-1_i386.deb
[sdl2trs_1.2.25-1_i386.deb]: bin/sdl2trs_1.2.25-1_i386.deb
[sdltrs_1.2.25-1_amd64.deb]: bin/sdltrs_1.2.25-1_amd64.deb
[sdl2trs_1.2.25-1_amd64.deb]: bin/sdl2trs_1.2.25-1_amd64.deb
[xtrs]: https://www.tim-mann.org/xtrs.html


CHANGELOG.md
## Version 1.2.25 Release 12-18-2022:

 * Added Prologica CP-500/M80 support by Leonardo Brondani Schenkel.
 * Added partial Aster CT-80 emulation.
 * Fixed Tandy/RS Double-Density Adapter Kit.
 * Fixed Warm Boot for TRS-80 Model III.

## Version 1.2.24 Release 07-24-2022:

 * Added 6845 CRTC Interlace Mode Emulation.
 * Added VideoExtension HRG for EACA EG 3200 Genie III.
 * Added Genieplus Memory Card for EACA EG 3200 Genie III.
 * Added character set for Meritum I (Mera-Elzab).
 * Fixed Virtual Keyboard for SDL2.
 * Fixed more bugs and other improvements.

## Version 1.2.23 Release 03-27-2022:

 * Added Anitek MegaMem card for TRS-80 Model III and 4/4P.
 * Added support for TRS-Model III 4 MHz speedup hardware.
 * Added SIO (serial port) for TCS Genie IIIs.
 * Added TCS Genie IIs/SpeedMaster RAM 192 B memory expansion.
 * Fixed TCS Genie IIs/SpeedMaster emulation.
 * Fixed boot of G-DOS 2.4 for Genie I/II.

## Version 1.2.22 Release 02-06-2022:

 * Added EG 64.1 memory expansion for TRS-80 Model I.
 * Added Schmidtke 80-Z Video Card for TRS-80 Model I.
 * Added EG 3210 Programmable Graphics Adaptor for Genie III.
 * Added some David Keil's TRS-80 Emulator extensions.
 * Added P2 key for TCS Genie IIs/SpeedMaster.
 * Fixed TRS-80 Selector mode 6.

## Version 1.2.21 Release 11-28-2021:

 * Added emulation of TCS Genie IIIs system.
 * Added support for Lubomir Soft Banker.
 * Added key bindings to switch Z80 CPU clock speed.

## Version 1.2.20 Release 07-31-2021:

  * Added 480x192 HRG resolution for LNW80 and TCS SpeedMaster/Genie IIs.
  * Added support for EG-64 Memory-Banking-Adaptor from TCS.
  * Added CP/M banking support for some TRS-80 Model I clones.
  * Fixed scaling of HRG1B graphics page.
  * Fixed Control Key for TRS-80 Model I.

## Version 1.2.19 Release 06-06-2021:

  * Added support for EACA EG 3200 (Genie III).

## Version 1.2.18 Release 05-09-2021:

  * Added Load/Save TRS-80 memory in debugger.
  * Added UART ports for System 80.
  * Added Model I ROM patch for auto-boot from hard drive.
  * Added support for hard disk images created with David Keil's emulator.
  * Fixed issue #7 "Level1 basic does not run" (reported by David Rand).

## Version 1.2.17 Release 04-04-2021:

  * Added support for Prologica CP-300/500 16kB ROM.
  * Added support for Seatronics Super Speed-Up Board.
  * Added SuperMem RAM and Prologica ROM to state file.

## Version 1.2.16 Release 02-28-2021:

  * Added joystick hat support in GUI.
  * Fixed HRG1B graphics page (pointed out by Stefano Bodrato).
  * Fixed fullscreen toggling for SDL2 Software Renderer.
  * Fixed bugs and other improvements.

## Version 1.2.15 Release 01-24-2021:

  * Added joystick emulation with mouse.
  * Fixed directory selection in GUI.
  * Fixed joystick function mapping.
  * Fixed some bugs and other improvements.

## Version 1.2.14 Release 11-29-2020:

  * Added brightness control for Scanlines.
  * Added shortcuts for: `-borderwidth`, `-cassette` and `-fullscreen`.
  * Fixed `LD A,R` instruction (pointed out by Timothy P. Mann).
  * Merged some changes for the zbx debugger from `xtrs`.

## Version 1.2.13 Release 09-27-2020:

  * Fixed issue #6 "Core dump when trying to print" (reported by "zoof").
  * Fixed loading of MODELA/III file.
  * Fixed TextGUI colors on big endian machines.
  * Fixed location of default directories.
  * Fixed some bugs and other improvements.

## Version 1.2.12 Release 08-16-2020:

  * Fixed issue #3 "Handle CapsLock key in SDL2" (reported by David Barr).
  * Fixed some crashes (Copy Buffer, "SuperMem" option).
  * Fixed bugs and other improvements.
  * Made Z80 debugger an optional feature.

## Version 1.2.11 Release 06-28-2020:

  * Added CPU clock speed selection.
  * Added R register to state file.
  * Fixed issue #2 "Spurious return characters" (reported by David Barr).
  * Internal improvements.

## Version 1.2.10 Release 05-24-2020:

  * Added Z80 refresh register.
  * Added more accurate Z80 block moves.
  * Added support for Sprinter II/III.
  * Fixed bugs and other improvements.

## Version 1.2.9a Release 04-19-2020:

  * Fixed crash in parity table on 64-bit systems.
  * Fixed crash in SDL cleanup.
  * Fixed crash/garbaled TRS-80 graphics.
  * Fixed mouse selection on window redraw.
  * Toggle joystick emulation with NumLock key.

## Version 1.2.9 Release 04-12-2020:

  * Fixed SDL2 regressions in version 1.2.8.
  * Added display of Z80 speed in window title bar.
  * Added "Turbo Paste" function.
  * Adjust Z80 clock rate on the fly.
  * Fixed bugs and other improvements.

## Version 1.2.8 Release 03-22-2020:

  * Major bugfix release.

## Version 1.2.7 Release 01-25-2020:

  * Fixed bugs and some small improvements.

## Version 1.2.6 Release 12-29-2019:

  * Added CPU panel to show Z80 registers in window title bar.
  * Fixed SDL2 keyboard handling and other bugs.

## Version 1.2.5 Release 12-21-2019:

  * Added function to select and execute CMD files directly.
  * Added function to save screenshot as BMP file.
  * Fixed various bugs.

## Version 1.2.4 Release 11-24-2019:

  * Added function to load and execute Model I CMD files directly.
  * Fixed some bugs.

## Version 1.2.3 Release 10-15-2019:

  * Added character sets for HT-1080Z, Video Genie/System 80/PMC-80.
  * Added support for function keys on Genie II/System 80 Mk II/PMC-81.
  * Added write protection for cassette images.
  * Fixed more bugs.

## Version 1.2.2 Release 08-18-2019:

  * Added support for Exatron Stringy Floppy.
  * Added new key bindings and menu options.
  * Fixed various bugs.

## Version 1.2.1 Release 07-24-2019:

  * Added LED for Turbo Mode.
  * Fixed various bugs.
  * SDL2 port should be stable.

## Version 1.2.0 Release 06-09-2019:

  * Included support for banked memory and Lowe LE18 by Alan Cox.
  * Fixed SegFaults: ROM Selection Menu, Scaling in Fullscreen.
  * New shortcut keys, key bindings and help screen.
  * Ctrl-A, Ctrl-C & Ctrl-V can now be used in the Emulator.

src/reed.h
/* Matthew Reed's hard drive format.  Thanks to Matthew for providing
   documentation.  The comments below are copied from his mail
   messages, with some additions. */

typedef struct {
  Uint8 id1;       /* 0: Identifier #1: 56H */
  Uint8 id2;       /* 1: Identifier #2: CBH */
  Uint8 ver;       /* 2: Version of format: 10H = version 1.0 */
  Uint8 cksum;     /* 3: Simple checksum:
		      To calculate, add together bytes 0 to 31 of header
		      (excepting byte 3), then XOR result with 4CH */
  Uint8 blks;      /* 4: Number of 256 byte blocks in header: should be 1 */
  Uint8 mb4;       /* 5: Not used, but HDFORMAT sets to 4 */
  Uint8 media;     /* 6: Media type: 0 for hard disk */
  Uint8 flag1;     /* 7: Flags #1:
		      bit 7: Write protected: 0 for no, 1 for yes
                             [xtrshard/dct ignores for now]
		      bit 6: Must be 0
		      bit 5 - 0: reserved */
  Uint8 flag2;     /* 8: Flags #2: reserved */
  Uint8 flag3;     /* 9: Flags #3: reserved */
  Uint8 crtr;      /* 10: Created by:
		      14H = HDFORMAT
		      42H = xtrs mkdisk
                      80H = Cervasio xtrshard port to Vavasour M4 emulator */
  Uint8 dfmt;      /* 11: Disk format: 0 = LDOS/LS-DOS */
  Uint8 mm;        /* 12: Creation month: mm */
  Uint8 dd;        /* 13: Creation day: dd */
  Uint8 yy;        /* 14: Creation year: yy (offset from 1900) */
  Uint8 res1[12];  /* 15 - 26: reserved */
  Uint8 dparm;     /* 27: Disk parameters: (unused with hard drives)
		      bit 7: Density: 0 = double, 1 = single
		      bit 6: Sides: 0 = one side, 1 = 2 sides
		      bit 5: First sector: 0 if sector 0, 1 if sector 1
		      bit 4: DAM convention: 0 if normal (LDOS),
		      1 if reversed (TRSDOS 1.3)
		      bit 3 - 0: reserved */
  Uint8 cyl;       /* 28: Number of cylinders per disk */
  Uint8 sec;       /* 29: Number of sectors per track (floppy); cyl (hard) */
  Uint8 gran;      /* 30: Number of granules per track (floppy); cyl (hard)*/
  Uint8 dcyl;      /* 31: Directory cylinder [mkdisk sets to 1; xtrs ignores]*/
  char label[32];  /* 32: Volume label: 31 bytes terminated by 0 */
  Uint8 res2[192]; /* 64 - 255: reserved */
} ReedHardHeader;
